In [2]:
#importing dependencies
import pandas as pd
from pathlib import Path
import numpy as np

In [3]:
#Setup
school_csv = Path("Resources/schools_complete.csv")
student_csv = Path("Resources/students_complete.csv")

#read the CSV files 
school_df = pd.read_csv(school_csv)
student_df = pd.read_csv(student_csv)

student_df

Student ID       student_name gender grade         school_name  \
0               0       Paul Bradley      M   9th   Huang High School   
1               1       Victor Smith      M  12th   Huang High School   
2               2    Kevin Rodriguez      M  12th   Huang High School   
3               3  Dr. Richard Scott      M  12th   Huang High School   
4               4         Bonnie Ray      F   9th   Huang High School   
...           ...                ...    ...   ...                 ...   
39165       39165       Donna Howard      F  12th  Thomas High School   
39166       39166          Dawn Bell      F  10th  Thomas High School   
39167       39167     Rebecca Tanner      F   9th  Thomas High School   
39168       39168       Desiree Kidd      F  10th  Thomas High School   
39169       39169    Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  
0                 66          79  
1                 94          61  
2                 90          60  
3                 67          58  
4                 97          84  
...              ...         ...  
39165             99          90  
39166             95          70  
39167             73          84  
39168             99          90  
39169             95          75  

[39170 rows x 7 columns]

In [4]:
school_df

School ID            school_name      type  size   budget
0           0      Huang High School  District  2917  1910635
1           1   Figueroa High School  District  2949  1884411
2           2    Shelton High School   Charter  1761  1056600
3           3  Hernandez High School  District  4635  3022020
4           4    Griffin High School   Charter  1468   917500
5           5     Wilson High School   Charter  2283  1319574
6           6    Cabrera High School   Charter  1858  1081356
7           7     Bailey High School  District  4976  3124928
8           8     Holden High School   Charter   427   248087
9           9       Pena High School   Charter   962   585858
10         10     Wright High School   Charter  1800  1049400
11         11  Rodriguez High School  District  3999  2547363
12         12    Johnson High School  District  4761  3094650
13         13       Ford High School  District  2739  1763916
14         14     Thomas High School   Charter  1635  1043130

In [5]:
#Merge the two dataframes into 1 dataframe 
combined_df = pd.merge(student_df, school_df, on="school_name", how='outer')
combined_df

Student ID       student_name gender grade         school_name  \
0               0       Paul Bradley      M   9th   Huang High School   
1               1       Victor Smith      M  12th   Huang High School   
2               2    Kevin Rodriguez      M  12th   Huang High School   
3               3  Dr. Richard Scott      M  12th   Huang High School   
4               4         Bonnie Ray      F   9th   Huang High School   
...           ...                ...    ...   ...                 ...   
39165       39165       Donna Howard      F  12th  Thomas High School   
39166       39166          Dawn Bell      F  10th  Thomas High School   
39167       39167     Rebecca Tanner      F   9th  Thomas High School   
39168       39168       Desiree Kidd      F  10th  Thomas High School   
39169       39169    Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
0                 66          79          0  District  2917  1910635  
1                 94          61          0  District  2917  1910635  
2                 90          60          0  District  2917  1910635  
3                 67          58          0  District  2917  1910635  
4                 97          84          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39165             99          90         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[39170 rows x 11 columns]

## District Summary

In [55]:
#Calculate total number of unique schools
unique_schools = combined_df["school_name"].unique()
total_schools = len(unique_schools)

total_schools

15

In [7]:
#Calculate the total number of students

total_students = combined_df["student_name"].count()

total_students

39170

In [8]:
#Calculate the total budget
#Need to look at the budget column in the school dataframe since you know what the budget allocation is per school

total_budget = school_df["budget"].sum()
total_budget

24649428

In [9]:
#Calculate the average math score
average_math_score = combined_df["math_score"].mean()
average_math_score

78.98537145774827

In [10]:
#Calculate the average reading score
average_reading_score = combined_df["reading_score"].mean()
average_reading_score

81.87784018381414

In [11]:
students_passed_math = combined_df[combined_df["math_score"] >= 70]
students_passed_math

Student ID     student_name gender grade         school_name  \
0               0     Paul Bradley      M   9th   Huang High School   
4               4       Bonnie Ray      F   9th   Huang High School   
5               5    Bryan Miranda      M   9th   Huang High School   
6               6    Sheena Carter      F  11th   Huang High School   
8               8     Michael Roth      M  10th   Huang High School   
...           ...              ...    ...   ...                 ...   
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
0                 66          79          0  District  2917  1910635  
4                 97          84          0  District  2917  1910635  
5                 94          94          0  District  2917  1910635  
6                 82          80          0  District  2917  1910635  
8                 95          87          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39165             99          90         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[29370 rows x 11 columns]

In [12]:
number_students_passed_math = students_passed_math["math_score"].count()
number_students_passed_math

29370

In [13]:
# Use the following to calculate the percentage of students who passed math (math scores greather than or equal to 70)
students_passed = combined_df[combined_df["math_score"] >= 70]
number_students_passed = students_passed["math_score"].count()
percentage_passed_math = (number_students_passed/total_students) * 100
percentage_passed_math

74.9808526933878

In [14]:
students_passed_reading = combined_df[combined_df["reading_score"] >= 70]
students_passed_reading

Student ID     student_name gender grade         school_name  \
1               1     Victor Smith      M  12th   Huang High School   
2               2  Kevin Rodriguez      M  12th   Huang High School   
4               4       Bonnie Ray      F   9th   Huang High School   
5               5    Bryan Miranda      M   9th   Huang High School   
6               6    Sheena Carter      F  11th   Huang High School   
...           ...              ...    ...   ...                 ...   
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
1                 94          61          0  District  2917  1910635  
2                 90          60          0  District  2917  1910635  
4                 97          84          0  District  2917  1910635  
5                 94          94          0  District  2917  1910635  
6                 82          80          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39165             99          90         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[33610 rows x 11 columns]

In [15]:
number_students_passed_reading = students_passed_reading["reading_score"].count()
number_students_passed_reading

33610

In [16]:
# Use the following to calculate the percentage of students who passed reading (reading scores greather than or equal to 70)
students_passed_reading = combined_df[combined_df["reading_score"] >= 70]
number_students_passed_reading = students_passed_reading["reading_score"].count()
percentage_passed_reading = (number_students_passed_reading/total_students) * 100
percentage_passed_reading

85.80546336482001

In [17]:
# Use the following to calculate the percentage of students that passed math and reading
passing_math_reading_count = combined_df[(combined_df["math_score"] >= 70) & (combined_df["reading_score"] >= 70)].count()["student_name"]
overall_passing_rate = passing_math_reading_count /  float(total_students) * 100
overall_passing_rate

65.17232575950983

In [18]:
# Create a high-level snapshot of the district's key metrics in a DataFrame
district_summary = pd.DataFrame({"Total Schools": [total_schools], "Total Students": [total_students], 
                                "Total Budget": [total_budget], "Average Math Score": [average_math_score],
                                "Average Reading Score": [average_reading_score], "% Passing Math": [percentage_passed_math],
                                "% Passing Reading": [percentage_passed_reading], "Overall Passing Rate": [overall_passing_rate]})

# Formatting
district_summary["Total Students"] = district_summary["Total Students"].map("{:,}".format)
district_summary["Total Budget"] = district_summary["Total Budget"].map("${:,.2f}".format)

# Display the DataFrame
district_summary

Total Schools Total Students    Total Budget  Average Math Score  \
0             15         39,170  $24,649,428.00           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  \
0               81.87784       74.980853          85.805463   

   Overall Passing Rate  
0             65.172326

## School Summary

In [19]:
#Use the code provided to select all of the school types
school_types = school_df.loc[:,["school_name","type"]]

school_types

school_name      type
0       Huang High School  District
1    Figueroa High School  District
2     Shelton High School   Charter
3   Hernandez High School  District
4     Griffin High School   Charter
5      Wilson High School   Charter
6     Cabrera High School   Charter
7      Bailey High School  District
8      Holden High School   Charter
9        Pena High School   Charter
10     Wright High School   Charter
11  Rodriguez High School  District
12    Johnson High School  District
13       Ford High School  District
14     Thomas High School   Charter

In [20]:
#Calculate the total student count per school

per_school_counts = school_df.loc[:,["school_name","size"]]

total_students_per_school = per_school_counts.rename(columns={"size": "Total Students"})

total_students_per_school

school_name  Total Students
0       Huang High School            2917
1    Figueroa High School            2949
2     Shelton High School            1761
3   Hernandez High School            4635
4     Griffin High School            1468
5      Wilson High School            2283
6     Cabrera High School            1858
7      Bailey High School            4976
8      Holden High School             427
9        Pena High School             962
10     Wright High School            1800
11  Rodriguez High School            3999
12    Johnson High School            4761
13       Ford High School            2739
14     Thomas High School            1635

In [21]:
#Calculate the total budget per school

per_school_budget = school_df.loc[:,["school_name","size","budget"]]

total_budget = per_school_budget.rename(columns = {"size":"Total Students","budget":"Total Budget"})

total_budget = total_budget.sort_values("school_name")

total_budget

school_name  Total Students  Total Budget
7      Bailey High School            4976       3124928
6     Cabrera High School            1858       1081356
1    Figueroa High School            2949       1884411
13       Ford High School            2739       1763916
4     Griffin High School            1468        917500
3   Hernandez High School            4635       3022020
8      Holden High School             427        248087
0       Huang High School            2917       1910635
12    Johnson High School            4761       3094650
9        Pena High School             962        585858
11  Rodriguez High School            3999       2547363
2     Shelton High School            1761       1056600
14     Thomas High School            1635       1043130
5      Wilson High School            2283       1319574
10     Wright High School            1800       1049400

In [22]:
# Calculate the total per capita spending per school

total_budget["Per Student Budget"] = (total_budget["Total Budget"] / total_budget["Total Students"])

total_budget

school_name  Total Students  Total Budget  Per Student Budget
7      Bailey High School            4976       3124928               628.0
6     Cabrera High School            1858       1081356               582.0
1    Figueroa High School            2949       1884411               639.0
13       Ford High School            2739       1763916               644.0
4     Griffin High School            1468        917500               625.0
3   Hernandez High School            4635       3022020               652.0
8      Holden High School             427        248087               581.0
0       Huang High School            2917       1910635               655.0
12    Johnson High School            4761       3094650               650.0
9        Pena High School             962        585858               609.0
11  Rodriguez High School            3999       2547363               637.0
2     Shelton High School            1761       1056600               600.0
14     Thomas High School            1635       1043130               638.0
5      Wilson High School            2283       1319574               578.0
10     Wright High School            1800       1049400               583.0

In [23]:
#First we need to create a dataframe that has just the school name, reading score and math score by schools

math_reading_score_group = combined_df.loc[:,["school_name","reading_score","math_score"]]

math_reading_score_group

school_name  reading_score  math_score
0       Huang High School             66          79
1       Huang High School             94          61
2       Huang High School             90          60
3       Huang High School             67          58
4       Huang High School             97          84
...                   ...            ...         ...
39165  Thomas High School             99          90
39166  Thomas High School             95          70
39167  Thomas High School             73          84
39168  Thomas High School             99          90
39169  Thomas High School             95          75

[39170 rows x 3 columns]

In [24]:
# Calculate the average test scores per school
per_school = math_reading_score_group.groupby(["school_name"]).mean()

average_per_school = per_school.rename(columns = {"reading_score": "Average Reading Score","math_score": "Average Math Score"})

average_per_school


Average Reading Score  Average Math Score
school_name                                                     
Bailey High School                 81.033963           77.048432
Cabrera High School                83.975780           83.061895
Figueroa High School               81.158020           76.711767
Ford High School                   80.746258           77.102592
Griffin High School                83.816757           83.351499
Hernandez High School              80.934412           77.289752
Holden High School                 83.814988           83.803279
Huang High School                  81.182722           76.629414
Johnson High School                80.966394           77.072464
Pena High School                   84.044699           83.839917
Rodriguez High School              80.744686           76.842711
Shelton High School                83.725724           83.359455
Thomas High School                 83.848930           83.418349
Wilson High School                 83.989488           83.274201
Wright High School                 83.955000           83.682222

In [25]:
# Calculate the number of students per school with math scores of 70 or higher

math_group = combined_df.loc[:,["school_name","math_score"]]

students_passing_math = math_group[math_group["math_score"] >= 70]

school_students_passing_math = students_passing_math.groupby(["school_name"]).count()

school_students_passing_math = school_students_passing_math.rename(columns = {"math_score": "Total Students who Passed Math"})

school_students_passing_math

Total Students who Passed Math
school_name                                          
Bailey High School                               3318
Cabrera High School                              1749
Figueroa High School                             1946
Ford High School                                 1871
Griffin High School                              1371
Hernandez High School                            3094
Holden High School                                395
Huang High School                                1916
Johnson High School                              3145
Pena High School                                  910
Rodriguez High School                            2654
Shelton High School                              1653
Thomas High School                               1525
Wilson High School                               2143
Wright High School                               1680

In [26]:
# Calculate the number of students per school with reading scores of 70 or higher

reading_group = combined_df.loc[:,["school_name","reading_score"]]

students_passing_reading = reading_group[reading_group["reading_score"] >= 70]

school_students_passing_reading = students_passing_reading.groupby(["school_name"]).count()

school_students_passing_reading = school_students_passing_reading.rename(columns = {"reading_score": "Total Students who Passed Reading"})

school_students_passing_reading

Total Students who Passed Reading
school_name                                             
Bailey High School                                  4077
Cabrera High School                                 1803
Figueroa High School                                2381
Ford High School                                    2172
Griffin High School                                 1426
Hernandez High School                               3748
Holden High School                                   411
Huang High School                                   2372
Johnson High School                                 3867
Pena High School                                     923
Rodriguez High School                               3208
Shelton High School                                 1688
Thomas High School                                  1591
Wilson High School                                  2204
Wright High School                                  1739

In [27]:
# Use the provided code to calculate the number of students per school that passed both math and reading with scores of 70 or higher

students_passing_math_and_reading = math_reading_score_group[(math_reading_score_group["reading_score"] >= 70) & (math_reading_score_group["math_score"] >= 70)]


school_students_passing_math_and_reading = students_passing_math_and_reading.groupby(["school_name"]).size()

school_students_passing_math_and_reading

school_name
Bailey High School       2719
Cabrera High School      1697
Figueroa High School     1569
Ford High School         1487
Griffin High School      1330
Hernandez High School    2481
Holden High School        381
Huang High School        1561
Johnson High School      2549
Pena High School          871
Rodriguez High School    2119
Shelton High School      1583
Thomas High School       1487
Wilson High School       2068
Wright High School       1626
dtype: int64

In [28]:
# DataFrame to show number of students who passed both math and reading
test_df = math_reading_score_group[(math_reading_score_group["reading_score"] >= 70) & (math_reading_score_group["math_score"] >= 70)]

#By using count, this returns a format in a dataframe that I can call
passing_math_and_reading_df = test_df.groupby(["school_name"]).count()

passing_math_and_reading_df = passing_math_and_reading_df.rename(columns = {"reading_score":"# of Students who passed Reading and Math"})

del passing_math_and_reading_df["math_score"]

passing_math_and_reading_df


# of Students who passed Reading and Math
school_name                                                     
Bailey High School                                          2719
Cabrera High School                                         1697
Figueroa High School                                        1569
Ford High School                                            1487
Griffin High School                                         1330
Hernandez High School                                       2481
Holden High School                                           381
Huang High School                                           1561
Johnson High School                                         2549
Pena High School                                             871
Rodriguez High School                                       2119
Shelton High School                                         1583
Thomas High School                                          1487
Wilson High School                                          2068
Wright High School                                          1626

In [29]:
# Merging the dataframes that contain the total students who passed math and the total students who passed reading per school

new_df = pd.merge(school_students_passing_math, school_students_passing_reading, on="school_name", how='outer')

new2_df = pd.merge(total_students_per_school, new_df, on="school_name", how='outer')

new3_df = pd.merge(passing_math_and_reading_df, new2_df, on="school_name", how='outer')

new3_df = new3_df.sort_values("school_name")

new3_df

school_name  # of Students who passed Reading and Math  \
0      Bailey High School                                       2719   
1     Cabrera High School                                       1697   
2    Figueroa High School                                       1569   
3        Ford High School                                       1487   
4     Griffin High School                                       1330   
5   Hernandez High School                                       2481   
6      Holden High School                                        381   
7       Huang High School                                       1561   
8     Johnson High School                                       2549   
9        Pena High School                                        871   
10  Rodriguez High School                                       2119   
11    Shelton High School                                       1583   
12     Thomas High School                                       1487   
13     Wilson High School                                       2068   
14     Wright High School                                       1626   

    Total Students  Total Students who Passed Math  \
0             4976                            3318   
1             1858                            1749   
2             2949                            1946   
3             2739                            1871   
4             1468                            1371   
5             4635                            3094   
6              427                             395   
7             2917                            1916   
8             4761                            3145   
9              962                             910   
10            3999                            2654   
11            1761                            1653   
12            1635                            1525   
13            2283                            2143   
14            1800                            1680   

    Total Students who Passed Reading  
0                                4077  
1                                1803  
2                                2381  
3                                2172  
4                                1426  
5                                3748  
6                                 411  
7                                2372  
8                                3867  
9                                 923  
10                               3208  
11                               1688  
12                               1591  
13                               2204  
14                               1739

In [30]:
# Use the provided code to calculate the passing rates per school
new3_df["% Passing Math"] = (new3_df["Total Students who Passed Math"] / new3_df["Total Students"]) * 100

new3_df["% Passing Reading"] = (new3_df["Total Students who Passed Reading"] / new3_df["Total Students"]) * 100


new3_df["% Overall Passing"] = (new3_df["# of Students who passed Reading and Math"] / new3_df["Total Students"]) * 100


In [31]:
new3_df

school_name  # of Students who passed Reading and Math  \
0      Bailey High School                                       2719   
1     Cabrera High School                                       1697   
2    Figueroa High School                                       1569   
3        Ford High School                                       1487   
4     Griffin High School                                       1330   
5   Hernandez High School                                       2481   
6      Holden High School                                        381   
7       Huang High School                                       1561   
8     Johnson High School                                       2549   
9        Pena High School                                        871   
10  Rodriguez High School                                       2119   
11    Shelton High School                                       1583   
12     Thomas High School                                       1487   
13     Wilson High School                                       2068   
14     Wright High School                                       1626   

    Total Students  Total Students who Passed Math  \
0             4976                            3318   
1             1858                            1749   
2             2949                            1946   
3             2739                            1871   
4             1468                            1371   
5             4635                            3094   
6              427                             395   
7             2917                            1916   
8             4761                            3145   
9              962                             910   
10            3999                            2654   
11            1761                            1653   
12            1635                            1525   
13            2283                            2143   
14            1800                            1680   

    Total Students who Passed Reading  % Passing Math  % Passing Reading  \
0                                4077       66.680064          81.933280   
1                                1803       94.133477          97.039828   
2                                2381       65.988471          80.739234   
3                                2172       68.309602          79.299014   
4                                1426       93.392371          97.138965   
5                                3748       66.752967          80.862999   
6                                 411       92.505855          96.252927   
7                                2372       65.683922          81.316421   
8                                3867       66.057551          81.222432   
9                                 923       94.594595          95.945946   
10                               3208       66.366592          80.220055   
11                               1688       93.867121          95.854628   
12                               1591       93.272171          97.308869   
13                               2204       93.867718          96.539641   
14                               1739       93.333333          96.611111   

    % Overall Passing  
0           54.642283  
1           91.334769  
2           53.204476  
3           54.289887  
4           90.599455  
5           53.527508  
6           89.227166  
7           53.513884  
8           53.539172  
9           90.540541  
10          52.988247  
11          89.892107  
12          90.948012  
13          90.582567  
14          90.333333

In [32]:
# Merging the different dataframes that I created to incorporate the final columns 

merge1_df = pd.merge(school_types,total_budget, on="school_name", how = "outer")

merge2_df = pd.merge(merge1_df, average_per_school, on="school_name", how = "outer")

merge3_df = pd.merge(merge2_df, new3_df, on="school_name", how = "outer")

del merge3_df["# of Students who passed Reading and Math"]
del merge3_df["Total Students_y"]
del merge3_df["Total Students who Passed Math"]
del merge3_df["Total Students who Passed Reading"]

merge3_df = merge3_df.rename(columns = {"type": "School Type","Total Students_x":"Total Students", "Total Budget":"Total School Budget"})

merge3_df = merge3_df.sort_values("school_name")


In [33]:
# Create a DataFrame called `per_school_summary` with columns for the calculations above.
per_school_summary = merge3_df.set_index("school_name")
per_school_summary.index.name = None

# Formatting
per_school_summary["Total School Budget"] = per_school_summary["Total School Budget"].map("${:,.2f}".format)
per_school_summary["Per Student Budget"] = per_school_summary["Per Student Budget"].map("${:,.2f}".format)

# Display the DataFrame
per_school_summary

School Type  Total Students Total School Budget  \
Bailey High School       District            4976       $3,124,928.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Figueroa High School     District            2949       $1,884,411.00   
Ford High School         District            2739       $1,763,916.00   
Griffin High School       Charter            1468         $917,500.00   
Hernandez High School    District            4635       $3,022,020.00   
Holden High School        Charter             427         $248,087.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Pena High School          Charter             962         $585,858.00   
Rodriguez High School    District            3999       $2,547,363.00   
Shelton High School       Charter            1761       $1,056,600.00   
Thomas High School        Charter            1635       $1,043,130.00   
Wilson High School        Charter            2283       $1,319,574.00   
Wright High School        Charter            1800       $1,049,400.00   

                      Per Student Budget  Average Reading Score  \
Bailey High School               $628.00              81.033963   
Cabrera High School              $582.00              83.975780   
Figueroa High School             $639.00              81.158020   
Ford High School                 $644.00              80.746258   
Griffin High School              $625.00              83.816757   
Hernandez High School            $652.00              80.934412   
Holden High School               $581.00              83.814988   
Huang High School                $655.00              81.182722   
Johnson High School              $650.00              80.966394   
Pena High School                 $609.00              84.044699   
Rodriguez High School            $637.00              80.744686   
Shelton High School              $600.00              83.725724   
Thomas High School               $638.00              83.848930   
Wilson High School               $578.00              83.989488   
Wright High School               $583.00              83.955000   

                       Average Math Score  % Passing Math  % Passing Reading  \
Bailey High School              77.048432       66.680064          81.933280   
Cabrera High School             83.061895       94.133477          97.039828   
Figueroa High School            76.711767       65.988471          80.739234   
Ford High School                77.102592       68.309602          79.299014   
Griffin High School             83.351499       93.392371          97.138965   
Hernandez High School           77.289752       66.752967          80.862999   
Holden High School              83.803279       92.505855          96.252927   
Huang High School               76.629414       65.683922          81.316421   
Johnson High School             77.072464       66.057551          81.222432   
Pena High School                83.839917       94.594595          95.945946   
Rodriguez High School           76.842711       66.366592          80.220055   
Shelton High School             83.359455       93.867121          95.854628   
Thomas High School              83.418349       93.272171          97.308869   
Wilson High School              83.274201       93.867718          96.539641   
Wright High School              83.682222       93.333333          96.611111   

                       % Overall Passing  
Bailey High School             54.642283  
Cabrera High School            91.334769  
Figueroa High School           53.204476  
Ford High School               54.289887  
Griffin High School            90.599455  
Hernandez High School          53.527508  
Holden High School             89.227166  
Huang High School              53.513884  
Johnson High School            53.539172  
Pena High School               90.540541  
Rodriguez High School     

## Highest-Performing Schools (by % Overall Passing)

In [34]:
# Sort the schools by `% Overall Passing` in descending order and display the top 5 rows.

top_schools = per_school_summary.sort_values("% Overall Passing", ascending=False)

top_schools.head(5)

School Type  Total Students Total School Budget  \
Cabrera High School     Charter            1858       $1,081,356.00   
Thomas High School      Charter            1635       $1,043,130.00   
Griffin High School     Charter            1468         $917,500.00   
Wilson High School      Charter            2283       $1,319,574.00   
Pena High School        Charter             962         $585,858.00   

                    Per Student Budget  Average Reading Score  \
Cabrera High School            $582.00              83.975780   
Thomas High School             $638.00              83.848930   
Griffin High School            $625.00              83.816757   
Wilson High School             $578.00              83.989488   
Pena High School               $609.00              84.044699   

                     Average Math Score  % Passing Math  % Passing Reading  \
Cabrera High School           83.061895       94.133477          97.039828   
Thomas High School            83.418349       93.272171          97.308869   
Griffin High School           83.351499       93.392371          97.138965   
Wilson High School            83.274201       93.867718          96.539641   
Pena High School              83.839917       94.594595          95.945946   

                     % Overall Passing  
Cabrera High School          91.334769  
Thomas High School           90.948012  
Griffin High School          90.599455  
Wilson High School           90.582567  
Pena High School             90.540541

## Bottom Performing Schools (By % Overall Passing)

In [35]:
# Sort the schools by `% Overall Passing` in ascending order and display the top 5 rows.

bottom_schools = per_school_summary.sort_values("% Overall Passing", ascending=True)
bottom_schools.head(5)

School Type  Total Students Total School Budget  \
Rodriguez High School    District            3999       $2,547,363.00   
Figueroa High School     District            2949       $1,884,411.00   
Huang High School        District            2917       $1,910,635.00   
Hernandez High School    District            4635       $3,022,020.00   
Johnson High School      District            4761       $3,094,650.00   

                      Per Student Budget  Average Reading Score  \
Rodriguez High School            $637.00              80.744686   
Figueroa High School             $639.00              81.158020   
Huang High School                $655.00              81.182722   
Hernandez High School            $652.00              80.934412   
Johnson High School              $650.00              80.966394   

                       Average Math Score  % Passing Math  % Passing Reading  \
Rodriguez High School           76.842711       66.366592          80.220055   
Figueroa High School            76.711767       65.988471          80.739234   
Huang High School               76.629414       65.683922          81.316421   
Hernandez High School           77.289752       66.752967          80.862999   
Johnson High School             77.072464       66.057551          81.222432   

                       % Overall Passing  
Rodriguez High School          52.988247  
Figueroa High School           53.204476  
Huang High School              53.513884  
Hernandez High School          53.527508  
Johnson High School            53.539172

## Math Scores by Grade

In [36]:
# Use the code provided to separate the data by grade
ninth_graders = combined_df[(combined_df["grade"] == "9th")]
tenth_graders = combined_df[(combined_df["grade"] == "10th")]
eleventh_graders = combined_df[(combined_df["grade"] == "11th")]
twelfth_graders = combined_df[(combined_df["grade"] == "12th")]


In [37]:
# Group by `school_name` and take the mean of the `math_score` column for each.
ninth_graders_math = ninth_graders.loc[:,["school_name","math_score"]]
ninth_grade_math_scores = ninth_graders_math.groupby("school_name").mean()
ninth_grade_math_scores = ninth_grade_math_scores.rename(columns={"math_score":"9th"})

tenth_graders_math = tenth_graders.loc[:,["school_name","math_score"]]
tenth_grade_math_scores = tenth_graders_math.groupby("school_name").mean()
tenth_grade_math_scores = tenth_grade_math_scores.rename(columns={"math_score":"10th"})

eleventh_graders_math = eleventh_graders.loc[:,["school_name","math_score"]]
eleventh_grade_math_scores = eleventh_graders_math.groupby("school_name").mean()
eleventh_grade_math_scores = eleventh_grade_math_scores.rename(columns={"math_score":"11th"})

twelfth_graders_math = twelfth_graders.loc[:,["school_name","math_score"]]
twelfth_grade_math_scores = twelfth_graders_math.groupby("school_name").mean()
twelfth_grade_math_scores = twelfth_grade_math_scores.rename(columns={"math_score":"12th"})

In [38]:
# Combine each of the scores above into single DataFrame called `math_scores_by_grade`

math1 = pd.merge(ninth_grade_math_scores, tenth_grade_math_scores, on="school_name", how="outer")

math2 = pd.merge(math1, eleventh_grade_math_scores, on="school_name", how="outer")

math_scores_by_grade = pd.merge(math2, twelfth_grade_math_scores, on="school_name", how="outer" )

In [39]:
# Minor data wrangling
math_scores_by_grade.index.name = None

# Display the DataFrame
math_scores_by_grade

9th       10th       11th       12th
Bailey High School     77.083676  76.996772  77.515588  76.492218
Cabrera High School    83.094697  83.154506  82.765560  83.277487
Figueroa High School   76.403037  76.539974  76.884344  77.151369
Ford High School       77.361345  77.672316  76.918058  76.179963
Griffin High School    82.044010  84.229064  83.842105  83.356164
Hernandez High School  77.438495  77.337408  77.136029  77.186567
Holden High School     83.787402  83.429825  85.000000  82.855422
Huang High School      77.027251  75.908735  76.446602  77.225641
Johnson High School    77.187857  76.691117  77.491653  76.863248
Pena High School       83.625455  83.372000  84.328125  84.121547
Rodriguez High School  76.859966  76.612500  76.395626  77.690748
Shelton High School    83.420755  82.917411  83.383495  83.778976
Thomas High School     83.590022  83.087886  83.498795  83.497041
Wilson High School     83.085578  83.724422  83.195326  83.035794
Wright High School     83.264706  84.010288  83.836782  83.644986

## Reading Score by Grade

In [40]:
# Use the code provided to separate the data by grade
ninth_graders = combined_df[(combined_df["grade"] == "9th")]
tenth_graders = combined_df[(combined_df["grade"] == "10th")]
eleventh_graders = combined_df[(combined_df["grade"] == "11th")]
twelfth_graders = combined_df[(combined_df["grade"] == "12th")]


In [41]:
# Group by `school_name` and take the mean of the `reading_score` column for each.
ninth_graders_reading = ninth_graders.loc[:,["school_name","reading_score"]]
ninth_grade_reading_scores = ninth_graders_reading.groupby("school_name").mean()
ninth_grade_reading_scores = ninth_grade_reading_scores.rename(columns={"reading_score":"9th"})

tenth_graders_reading = tenth_graders.loc[:,["school_name","reading_score"]]
tenth_grade_reading_scores = tenth_graders_reading.groupby("school_name").mean()
tenth_grade_reading_scores = tenth_grade_reading_scores.rename(columns={"reading_score":"10th"})

eleventh_graders_reading = eleventh_graders.loc[:,["school_name","reading_score"]]
eleventh_grade_reading_scores = eleventh_graders_reading.groupby("school_name").mean()
eleventh_grade_reading_scores = eleventh_grade_reading_scores.rename(columns={"reading_score":"11th"})

twelfth_graders_reading = twelfth_graders.loc[:,["school_name","reading_score"]]
twelfth_grade_reading_scores = twelfth_graders_reading.groupby("school_name").mean()
twelfth_grade_reading_scores = twelfth_grade_reading_scores.rename(columns={"reading_score":"12th"})

In [42]:
# Combine each of the scores above into single DataFrame called `reading_scores_by_grade`

reading1 = pd.merge(ninth_grade_reading_scores, tenth_grade_reading_scores, on="school_name", how="outer")

reading2 = pd.merge(reading1, eleventh_grade_reading_scores, on="school_name", how="outer")

reading_scores_by_grade = pd.merge(math2, twelfth_grade_reading_scores, on="school_name", how="outer" )


In [43]:
# Minor data wrangling
reading_scores_by_grade.index.name = None

# Display the DataFrame
reading_scores_by_grade

9th       10th       11th       12th
Bailey High School     77.083676  76.996772  77.515588  80.912451
Cabrera High School    83.094697  83.154506  82.765560  84.287958
Figueroa High School   76.403037  76.539974  76.884344  81.384863
Ford High School       77.361345  77.672316  76.918058  80.662338
Griffin High School    82.044010  84.229064  83.842105  84.013699
Hernandez High School  77.438495  77.337408  77.136029  80.857143
Holden High School     83.787402  83.429825  85.000000  84.698795
Huang High School      77.027251  75.908735  76.446602  80.305983
Johnson High School    77.187857  76.691117  77.491653  81.227564
Pena High School       83.625455  83.372000  84.328125  84.591160
Rodriguez High School  76.859966  76.612500  76.395626  80.376426
Shelton High School    83.420755  82.917411  83.383495  82.781671
Thomas High School     83.590022  83.087886  83.498795  83.831361
Wilson High School     83.085578  83.724422  83.195326  84.317673
Wright High School     83.264706  84.010288  83.836782  84.073171

## Scores by School Spending

In [44]:
# Establish the bins 
spending_bins = [0, 585, 630, 645, 680]

spending_labels = ["<$585", "$585-630", "$630-645", "$645-680"]

In [45]:
# Create a copy of the merge3_df since the values don't have dollar signs in the "Per Student Budget" column
school_spending_df = merge3_df.copy()
school_spending_df

school_name School Type  Total Students  Total School Budget  \
7      Bailey High School    District            4976              3124928   
6     Cabrera High School     Charter            1858              1081356   
1    Figueroa High School    District            2949              1884411   
13       Ford High School    District            2739              1763916   
4     Griffin High School     Charter            1468               917500   
3   Hernandez High School    District            4635              3022020   
8      Holden High School     Charter             427               248087   
0       Huang High School    District            2917              1910635   
12    Johnson High School    District            4761              3094650   
9        Pena High School     Charter             962               585858   
11  Rodriguez High School    District            3999              2547363   
2     Shelton High School     Charter            1761              1056600   
14     Thomas High School     Charter            1635              1043130   
5      Wilson High School     Charter            2283              1319574   
10     Wright High School     Charter            1800              1049400   

    Per Student Budget  Average Reading Score  Average Math Score  \
7                628.0              81.033963           77.048432   
6                582.0              83.975780           83.061895   
1                639.0              81.158020           76.711767   
13               644.0              80.746258           77.102592   
4                625.0              83.816757           83.351499   
3                652.0              80.934412           77.289752   
8                581.0              83.814988           83.803279   
0                655.0              81.182722           76.629414   
12               650.0              80.966394           77.072464   
9                609.0              84.044699           83.839917   
11               637.0              80.744686           76.842711   
2                600.0              83.725724           83.359455   
14               638.0              83.848930           83.418349   
5                578.0              83.989488           83.274201   
10               583.0              83.955000           83.682222   

    % Passing Math  % Passing Reading  % Overall Passing  
7        66.680064          81.933280          54.642283  
6        94.133477          97.039828          91.334769  
1        65.988471          80.739234          53.204476  
13       68.309602          79.299014          54.289887  
4        93.392371          97.138965          90.599455  
3        66.752967          80.862999          53.527508  
8        92.505855          96.252927          89.227166  
0        65.683922          81.316421          53.513884  
12       66.057551          81.222432          53.539172  
9        94.594595          95.945946          90.540541  
11       66.366592          80.220055          52.988247  
2        93.867121          95.854628          89.892107  
14       93.272171          97.308869          90.948012  
5        93.867718          96.539641          90.582567  
10       93.333333          96.611111          90.333333

In [46]:
# Use `pd.cut` to categorize spending based on the bins.
school_spending_df["Spending Ranges (Per Student)"] = pd.cut(school_spending_df["Per Student Budget"], 
                                              spending_bins, labels=spending_labels, 
                                              include_lowest=True)

school_spending_df

school_name School Type  Total Students  Total School Budget  \
7      Bailey High School    District            4976              3124928   
6     Cabrera High School     Charter            1858              1081356   
1    Figueroa High School    District            2949              1884411   
13       Ford High School    District            2739              1763916   
4     Griffin High School     Charter            1468               917500   
3   Hernandez High School    District            4635              3022020   
8      Holden High School     Charter             427               248087   
0       Huang High School    District            2917              1910635   
12    Johnson High School    District            4761              3094650   
9        Pena High School     Charter             962               585858   
11  Rodriguez High School    District            3999              2547363   
2     Shelton High School     Charter            1761              1056600   
14     Thomas High School     Charter            1635              1043130   
5      Wilson High School     Charter            2283              1319574   
10     Wright High School     Charter            1800              1049400   

    Per Student Budget  Average Reading Score  Average Math Score  \
7                628.0              81.033963           77.048432   
6                582.0              83.975780           83.061895   
1                639.0              81.158020           76.711767   
13               644.0              80.746258           77.102592   
4                625.0              83.816757           83.351499   
3                652.0              80.934412           77.289752   
8                581.0              83.814988           83.803279   
0                655.0              81.182722           76.629414   
12               650.0              80.966394           77.072464   
9                609.0              84.044699           83.839917   
11               637.0              80.744686           76.842711   
2                600.0              83.725724           83.359455   
14               638.0              83.848930           83.418349   
5                578.0              83.989488           83.274201   
10               583.0              83.955000           83.682222   

    % Passing Math  % Passing Reading  % Overall Passing  \
7        66.680064          81.933280          54.642283   
6        94.133477          97.039828          91.334769   
1        65.988471          80.739234          53.204476   
13       68.309602          79.299014          54.289887   
4        93.392371          97.138965          90.599455   
3        66.752967          80.862999          53.527508   
8        92.505855          96.252927          89.227166   
0        65.683922          81.316421          53.513884   
12       66.057551          81.222432          53.539172   
9        94.594595          95.945946          90.540541   
11       66.366592          80.220055          52.988247   
2        93.867121          95.854628          89.892107   
14       93.272171          97.308869          90.948012   
5        93.867718          96.539641          90.582567   
10       93.333333          96.611111          90.333333   

   Spending Ranges (Per Student)  
7                       $585-630  
6                          <$585  
1                       $630-645  
13                      $630-645  
4                       $585-630  
3                       $645-680  
8                          <$585  
0                       $645-680  
12                      $645-680  
9                       $585-630  
11                      $630-645  
2                       $585-630  
14                      $630-645  
5                          <$585  
10                         <$585

In [47]:
#  Calculate averages for the desired columns. 
spending_math_scores = school_spending_df.groupby(["Spending Ranges (Per Student)"])["Average Math Score"].mean()
spending_reading_scores = school_spending_df.groupby(["Spending Ranges (Per Student)"])["Average Reading Score"].mean()
spending_passing_math = school_spending_df.groupby(["Spending Ranges (Per Student)"])["% Passing Math"].mean()
spending_passing_reading = school_spending_df.groupby(["Spending Ranges (Per Student)"])["% Passing Reading"].mean()
overall_passing_spending = school_spending_df.groupby(["Spending Ranges (Per Student)"])["% Overall Passing"].mean()

In [48]:
# Assemble into DataFrame
spending_dict = {'Average Math Score': spending_math_scores,
                'Average Reading Score': spending_reading_scores,
                '% Passing Math': spending_passing_math,
                '% Passing Reading': spending_passing_reading,
                '% Overall Passing': overall_passing_spending}
spending_summary = pd.DataFrame(spending_dict)

# Display results
spending_summary

Average Math Score  Average Reading Score  \
Spending Ranges (Per Student)                                              
<$585                                   83.455399              83.933814   
$585-630                                81.899826              83.155286   
$630-645                                78.518855              81.624473   
$645-680                                76.997210              81.027843   

                               % Passing Math  % Passing Reading  \
Spending Ranges (Per Student)                                      
<$585                               93.460096          96.610877   
$585-630                            87.133538          92.718205   
$630-645                            73.484209          84.391793   
$645-680                            66.164813          81.133951   

                               % Overall Passing  
Spending Ranges (Per Student)                     
<$585                                  90.369459  
$585-630                               81.418596  
$630-645                               62.857656  
$645-680                               53.526855

## Scores by School Size

In [49]:
# Establish the bins.
size_bins = [0, 1000, 2000, 5000]
size_labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

In [50]:
# Categorize the spending based on the bins
# Use `pd.cut` on the "Total Students" column of the `per_school_summary` DataFrame.

per_school_summary["School Size"] = pd.cut(per_school_summary["Total Students"], 
                                              size_bins, labels=size_labels, 
                                              include_lowest=True)
per_school_summary

School Type  Total Students Total School Budget  \
Bailey High School       District            4976       $3,124,928.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Figueroa High School     District            2949       $1,884,411.00   
Ford High School         District            2739       $1,763,916.00   
Griffin High School       Charter            1468         $917,500.00   
Hernandez High School    District            4635       $3,022,020.00   
Holden High School        Charter             427         $248,087.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Pena High School          Charter             962         $585,858.00   
Rodriguez High School    District            3999       $2,547,363.00   
Shelton High School       Charter            1761       $1,056,600.00   
Thomas High School        Charter            1635       $1,043,130.00   
Wilson High School        Charter            2283       $1,319,574.00   
Wright High School        Charter            1800       $1,049,400.00   

                      Per Student Budget  Average Reading Score  \
Bailey High School               $628.00              81.033963   
Cabrera High School              $582.00              83.975780   
Figueroa High School             $639.00              81.158020   
Ford High School                 $644.00              80.746258   
Griffin High School              $625.00              83.816757   
Hernandez High School            $652.00              80.934412   
Holden High School               $581.00              83.814988   
Huang High School                $655.00              81.182722   
Johnson High School              $650.00              80.966394   
Pena High School                 $609.00              84.044699   
Rodriguez High School            $637.00              80.744686   
Shelton High School              $600.00              83.725724   
Thomas High School               $638.00              83.848930   
Wilson High School               $578.00              83.989488   
Wright High School               $583.00              83.955000   

                       Average Math Score  % Passing Math  % Passing Reading  \
Bailey High School              77.048432       66.680064          81.933280   
Cabrera High School             83.061895       94.133477          97.039828   
Figueroa High School            76.711767       65.988471          80.739234   
Ford High School                77.102592       68.309602          79.299014   
Griffin High School             83.351499       93.392371          97.138965   
Hernandez High School           77.289752       66.752967          80.862999   
Holden High School              83.803279       92.505855          96.252927   
Huang High School               76.629414       65.683922          81.316421   
Johnson High School             77.072464       66.057551          81.222432   
Pena High School                83.839917       94.594595          95.945946   
Rodriguez High School           76.842711       66.366592          80.220055   
Shelton High School             83.359455       93.867121          95.854628   
Thomas High School              83.418349       93.272171          97.308869   
Wilson High School              83.274201       93.867718          96.539641   
Wright High School              83.682222       93.333333          96.611111   

                       % Overall Passing         School Size  
Bailey High School             54.642283   Large (2000-5000)  
Cabrera High School            91.334769  Medium (1000-2000)  
Figueroa High School           53.204476   Large (2000-5000)  
Ford High School               54.289887   Large (2000-5000)  
Griffin High School            90.599455  Medium (1000-2000)  
Hernandez High School          53.527508   Large (2000-5000)  
Holden High School             89.227166       Small (<100

In [51]:
# Calculate averages for the desired columns. 
size_math_scores = per_school_summary.groupby(["School Size"])["Average Math Score"].mean()
size_reading_scores = per_school_summary.groupby(["School Size"])["Average Reading Score"].mean()
size_passing_math = per_school_summary.groupby(["School Size"])["% Passing Math"].mean()
size_passing_reading = per_school_summary.groupby(["School Size"])["% Passing Reading"].mean()
size_overall_passing = per_school_summary.groupby(["School Size"])["% Overall Passing"].mean()

In [52]:
# Assemble into DataFrame
size_dict = {'Average Math Score': size_math_scores,
                'Average Reading Score': size_reading_scores,
                '% Passing Math': size_passing_math,
                '% Passing Reading': size_passing_reading,
                '% Overall Passing': size_overall_passing}
size_summary = pd.DataFrame(size_dict)

# Display results
size_summary

Average Math Score  Average Reading Score  % Passing Math  \
School Size                                                                     
Small (<1000)                83.821598              83.929843       93.550225   
Medium (1000-2000)           83.374684              83.864438       93.599695   
Large (2000-5000)            77.746417              81.344493       69.963361   

                    % Passing Reading  % Overall Passing  
School Size                                               
Small (<1000)               96.099437          89.883853  
Medium (1000-2000)          96.790680          90.621535  
Large (2000-5000)           82.766634          58.286003

## Scores by School Type

In [53]:
# Group the per_school_summary DataFrame by "School Type" and average the results.

average_math_score_by_type = per_school_summary.groupby(["School Type"])["Average Math Score"].mean()
average_reading_score_by_type = per_school_summary.groupby(["School Type"])["Average Reading Score"].mean()
average_percent_passing_math_by_type = per_school_summary.groupby(["School Type"])["% Passing Math"].mean()
average_percent_passing_reading_by_type = per_school_summary.groupby(["School Type"])["% Passing Reading"].mean()
average_percent_overall_passing_by_type = per_school_summary.groupby(["School Type"])["% Overall Passing"].mean()

In [54]:
# Assemble into DataFrame
type_dict = {'Average Math Score': average_math_score_by_type,
                'Average Reading Score': average_reading_score_by_type,
                '% Passing Math': average_percent_passing_math_by_type,
                '% Passing Reading': average_percent_passing_reading_by_type,
                '% Overall Passing': average_percent_overall_passing_by_type}
type_summary = pd.DataFrame(type_dict)

# Display results
type_summary

Average Math Score  Average Reading Score  % Passing Math  \
School Type                                                              
Charter               83.473852              83.896421       93.620830   
District              76.956733              80.966636       66.548453   

             % Passing Reading  % Overall Passing  
School Type                                        
Charter              96.586489          90.432244  
District             80.799062          53.672208

## Analysis

### 
From the district_summary dataframe, it can be seen that across all 15 schools, the students on average performed better in reading that in math. The percentage of students who passed reading was 85.81 %, while the percentage of students who passed reading was 74.98%. To get useful information regarding the schools in the district, I calculated the total number of students, the total budget, the budget per student, the average math and reading scores for each school, and the percentage of students who passed math and reading. 

From the school_summary dataframe, it can be observed that the top 5 schools that had the greatest overall passing rates were charter schools and the schools that had the lowest overall passing rates were district schools. 

From the spending_summary, it can be seen that when the spending range is less than $585, the average math and reading scores and the percentage of students passing is greater than when the amount per student increases. 

From the size_summary, it can be seen the overall passing rate was greater for schools with less than 1000 students. Schools that had 2000-5000 students had a lower overall passing rate. 

From the type_summary dataframe, we can observe that the average math scores, reading scores, and the percentage of overall passing is greater for charter schools than district schools.  
